pkg/scheduler/dependency_graph.go

```diff
@@ -213,11 +261,137 @@ func NewDependencyGraph(sched *Scheduler, options interfaces.DependencyGraphOpti
+	context.args = make(map[string]string)
+	for key, value := range dependency.Args {
+		context.args[key] = copier.EvaluateString(value, parentContext.GetArg)
+	}
+	return context
+}
+
+func (sched *Scheduler) updateContext(context, parentContext *GraphContext, dependency client.Dependency) {
+	for key, value := range dependency.Args {
+		context.args[key] = copier.EvaluateString(value, parentContext.GetArg)
+	}
+}
+
+func (sched *Scheduler) newDefaultFlowObject() *client.Flow {
+	return &client.Flow{
+		TypeMeta: unversioned.TypeMeta{
+			Kind:       "Flow",
+			APIVersion: client.GroupName + "/" + client.Version,
+		},
+		ObjectMeta: api.ObjectMeta{
+			Name:      interfaces.DefaultFlowName,
+			Namespace: sched.client.Namespace(),
+		},
+		Exported:     true,
+	}
+}
+
+func checkArgs(options interfaces.DependencyGraphOptions, flow *client.Flow) error {
+	if !options.AllowUndeclaredArgs {
+		for key := range options.Args {
+			if _, ok := flow.Parameters[key]; !ok {
+				return fmt.Errorf("unexpected argument %s", key)
+			}
+		}
+	}
+	for key, value := range flow.Parameters {
+		if value.Default == nil {
+			if _, ok := options.Args[key]; !ok {
+				return fmt.Errorf("mandatory argument %s was not provided", key)
+			}
+		}
+	}
+	return nil
+}
+
+func unique(slice []string) []string {
+	seen := map[string]bool{}
+	var result []string
+	for _, t := range slice {
+		if _, found := seen[t]; !found {
+			result = append(result, t)
+			seen[t] = true
+		}
+	}
+	return result
+}
+
+type ReplicaList client.ReplicaList
+var _ sort.Interface = &ReplicaList{}
+
+func (rl *ReplicaList) Len() int {
+	return len(rl.Items)
+}
+
+func (rl *ReplicaList) Less(i, j int) bool {
+	if rl.Items[i].CreationTimestamp.Equal(rl.Items[j].CreationTimestamp) {
+		return rl.Items[i].UID < rl.Items[j].UID
+	}
+	return rl.Items[i].CreationTimestamp.Before(rl.Items[j].CreationTimestamp)
+}
+
+func (rl *ReplicaList) Swap(i, j int) {
+	rl.Items[i], rl.Items[j] = rl.Items[j], rl.Items[i]
+}
+
+func (sched *Scheduler) allocateReplicas(flow *client.Flow, options interfaces.DependencyGraphOptions) ([]string, error) {
+	label := labels.Set{"flow": flow.Name}
+	var tmp ReplicaList
+	var maxCurrentTime time.Time
+
+	count := options.ReplicaCount
+	if !options.ReplicaCountIsRelative || options.ReplicaCount == 0 {
+		existingReplicas, err := sched.client.Replicas().List(api.ListOptions{
+			LabelSelector: labels.SelectorFromSet(label),
+		})
+		if err != nil {
+			return nil, err
+		}
+		for _, item := range existingReplicas.Items {
+			if item.CreationTimestamp.After(maxCurrentTime) {
+				maxCurrentTime = item.CreationTimestamp.Time
+			}
+		}
+		tmp = ReplicaList(*existingReplicas)
+		if options.ReplicaCount == 0 {
+			count = len(tmp.Items)
+		}
+	}
+
+	for len(tmp.Items) < count {
+		replica := &client.Replica{}
+		replica.GenerateName = strings.ToLower(flow.Name) + "-"
+		replica.FlowName = flow.Name
+		replica.Labels = label
+		replica.Namespace = sched.client.Namespace()
```

Comment:
Why there are multiple assignments instead of single assign of struct with default values for struct members?